Features/cli app

simnh opened the pull request.

simnh committed:
Add script to build and compute package

simnh committed:
Pepify with black

simnh committed:
Update cli tools for compute and construct command

simnh committed:
Update commandline script

simnh committed:
Add option to construct datapackage

simnh committed:
Add option to construct datapackage

simnh committed:
Merge branch 'features/cli-app' of github.com:znes/fuchur into features/cli-app

simnh committed:
Adapt fuchur to postprocessing function

simnh committed:
Adapt fuchur to toml file

simnh committed:
Fix bugs from json to toml changes

simnh committed:
Fix scripts from restructuring and toml config

simnh committed:
Fix hydro model of reservoir

simnh committed:
Fix linter warnings

simnh committed:
Fix linter warnings

simnh committed:
Merge branch 'features/cli-app' of https://github.com/znes/fuchur into features/cli-app

simnh committed:
Merge branch 'features/cli-app' of https://github.com/znes/fuchur into features/cli-app

simnh committed:
Merge branch 'features/cli-app' of https://github.com/znes/fuchur into features/cli-app

simnh committed:
Fix data processing for GB country

simnh committed:
Update scenarios

simnh committed:
Add temporal resolution and emission limit to model stats

simnh committed:
Fix bugs for all countries with potentials and offshore

gnn commented on src/fuchur/__init__.py:
```diff
@@ -1 +1,7 @@
+import os
+
 __version__ = '0.0.0'
+__RAW_DATA_PATH__ = os.path.join(os.path.expanduser('~'), "fuchur-raw-data")
```
Is there a particular reason for hardcoding this and not making it configurable?

simnh replied:
No, just that I don't know how..

gnn commented on src/fuchur/cli.py:
```diff
@@ -14,10 +14,123 @@
+        Config dict for contructing the datapackage
+    """
+
+    datapackage.processing.clean_datapackage(
```
This and the function below are weird. The `_datapackage` suffix hint at the fact that we might want `datapackage.clean` and `datapackage.initialize`.
Also, `black` would want the closing parenthesis for `initialize`  on the next line. ;)

simnh replied:
ok i will rename to clean and initialize

gnn commented on src/fuchur/cli.py:
```diff
@@ -14,10 +14,123 @@
+        directories=["data", "resources"]
+    )
+
+    datapackage.building.initialize_datapackage(
```
All those calls having nearly the same signature could potentially be written shorter. But they also hint at a refactoring potential. Don't know whether any of this would really be an improvement though. The first option would be shorter, but harder to read. For refactoring I would have to check out the functions in question.

simnh replied:
I dont know what you mean...

gnn commented on src/fuchur/cli.py:
```diff
@@ -14,10 +14,123 @@
+    help="Protect results from being overwritten.",
+    )
+@click.pass_context
+def cli(ctx, solver, datapackage_dir, results_dir, temporal_resolution,
```
Why not just do:

```python
def cli(ctx, **kwargs):
    ctx.obj = {}
    ctx.obj["options"] = kwargs
```

Yes, you have to change all the keys to lowercase, but it's a small price to pay for being able to add options easier.
You could just do `ctx.obj = kwargs` but doing it like above makes the meaning of the dictionary entry explicit and frees the top level of `ctx.obj` up for storing other stuff too.

simnh replied:
done

gnn commented on src/fuchur/cli.py:
```diff
@@ -14,10 +14,123 @@
+@click.pass_context
+def construct(ctx, config):
+    config = datapackage.building.read_build_config(config)
+    _construct(config, ctx)
```
I'd avoid those underscore methods. If you don't want to clutter the `fuchur.cli` with these long functions, put them in another module.
Also, you don't have to import as `_compute` to avoid the name clash. If you just do `import fuchur.scripts.compute`, the module will only be available under its fully qualified name. ;)

simnh replied:
You can change this if you like

gnn commented on src/fuchur/scenarios/scenario2.toml:
```diff
@@ -0,0 +1,107 @@
+# Toml document to used for building model input
```
How about using those comments to explain the scenario? Then the two files will no longer have the same comment. :)

simnh replied:
yes but scenarios are still in progress

gnn commented on src/fuchur/scripts/__init__.py:
```diff
@@ -0,0 +1 @@
+__version__ = "0.0.0"
```
Drop the copied version number.

```suggestion
```

gnn commented on src/fuchur/scripts/bus.py:
```diff
@@ -0,0 +1,114 @@
+# -*- coding: utf-8 -*-
+"""
+This script constructs a pandas.Series `buses` with hub-names as index and
+polygons of these buses as values. It uses the NUTS shapefile.
```
Which NUTS shapefile?

gnn commented on src/fuchur/scripts/bus.py:
```diff
@@ -0,0 +1,114 @@
+
+    if not os.path.exists(filepath):
+        print("Shapefile data not found. Did you download raw data?")
+    # get nuts 1 regions for german neighbours
```
Erm... that formatting reads like the comment belongs to the line above.

gnn commented on src/fuchur/scripts/bus.py:
```diff
@@ -0,0 +1,114 @@
+
+    bio_potential = (
+        Package(
+            "https://raw.githubusercontent.com/ZNES-datapackages/technology-potential/master/datapackage.json"
```
Line to long. Probably needs to be wrapped.

```suggestion
            "https://raw.githubusercontent.com/ZNES-datapackages/"
            "technology-potential/master/datapackage.json"
```

gnn commented on src/fuchur/scripts/electricity.py:
```diff
@@ -0,0 +1,555 @@
+            parse_dates={"date": [0, 1, 2]},
+        ).set_index("date")["Inflow [GWh]"]
+
+    europe = [
```
Could this be moved to the configuration file?
No idea whether this makes sense, but its a bit long and the file already contains similar data.

simnh replied:
no dont think so

gnn commented on src/fuchur/scripts/electricity.py:
```diff
@@ -0,0 +1,555 @@
+            hyd.index.size
+        )  # normalize to new sampling frequency
+    # else:
+    #    normalization_factor = hydro.sum() / hyd.sum() #conserve total inflow for each country separately
```
```suggestion
    #     # conserve total inflow for each country separately
    #    normalization_factor = hydro.sum() / hyd.sum()
```

gnn commented on src/fuchur/scripts/electricity.py:
```diff
@@ -0,0 +1,555 @@
+
+    technologies = pd.DataFrame(
+        Package(
+            "https://raw.githubusercontent.com/ZNES-datapackages/technology-cost/master/datapackage.json"
```
```suggestion
            "https://raw.githubusercontent.com/ZNES-datapackages/"
            "technology-cost/master/datapackage.json"
```

gnn commented on src/fuchur/scripts/electricity.py:
```diff
@@ -0,0 +1,555 @@
+
+    potential = (
+        Package(
+            "https://raw.githubusercontent.com/ZNES-datapackages/technology-potential/master/datapackage.json"
```
```suggestion
            "https://raw.githubusercontent.com/ZNES-datapackages/"
            "technology-potential/master/datapackage.json"
```

gnn commented on src/fuchur/scripts/electricity.py:
```diff
@@ -0,0 +1,555 @@
+                                float(data["lifetime"]),
+                                wacc,
+                            )
+                            * 1000,  # €/kW -> €/MW
```
Best comment ever. Seriously. It's short and tells everybody something non-obvious they need to know. 👍

gnn commented on src/fuchur/scripts/electricity.py:
```diff
@@ -0,0 +1,555 @@
+
+    technologies = pd.DataFrame(
+        Package(
+            "https://raw.githubusercontent.com/ZNES-datapackages/technology-cost/master/datapackage.json"
```
```suggestion
            "https://raw.githubusercontent.com/ZNES-datapackages/"
            "technology-cost/master/datapackage.json"
```

gnn commented on src/fuchur/scripts/electricity.py:
```diff
@@ -0,0 +1,555 @@
+        .reset_index("carrier")
+        .apply(lambda x: dict({"carrier": x.carrier}, **x[0]), axis=1)
+    )
+    technologies = technologies.loc[config["temporal"]["scenario_year"]].to_dict()
```
```suggestion
    technologies = technologies.loc[
        config["temporal"]["scenario_year"]
    ].to_dict()
```

gnn reviewed (commented):
This is what I found so far. Only minor things, nothing big.
A few more general notes:

  * Maybe we can drop `download_rawdata.py` and move the single function call to the cli?
  * `scripts` is a rather general name that doesn't convey any meaning. How about moving the modules below it to the top or find places for them with more speaking names.
  * Is there a way to merge the two configuration files? Maybe by having a `scenario` section?

But `bus.py`, `capacity_factors.py`, `compute.py`, `electricity.py` and `grid.py` are rather long and not exactly easy to figure out. So I'll have to spend more time reading them to figure out whether there's a necessity to refactor them.

simnh committed:
Fix offhsore technologies

simnh committed:
Add heat script

simnh committed:
Merge branch 'features/cli-app' of github.com:znes/fuchur into features/cli-app

simnh committed:
Move Download data to cli script

simnh committed:
Fix test scenario

simnh committed:
Make cli options more flexible with kwargs

gnn committed:
Update src/fuchur/scripts/__init__.py

Co-Authored-By: simnh <simon.hilpert@uni-flensburg.de>

gnn committed:
Update src/fuchur/scripts/bus.py

Co-Authored-By: simnh <simon.hilpert@uni-flensburg.de>

gnn committed:
Update src/fuchur/scripts/electricity.py

Co-Authored-By: simnh <simon.hilpert@uni-flensburg.de>

gnn committed:
Update src/fuchur/scripts/electricity.py

Co-Authored-By: simnh <simon.hilpert@uni-flensburg.de>

gnn committed:
Update src/fuchur/scripts/electricity.py

Co-Authored-By: simnh <simon.hilpert@uni-flensburg.de>

gnn committed:
Update src/fuchur/scripts/electricity.py

Co-Authored-By: simnh <simon.hilpert@uni-flensburg.de>

simnh committed:
Fix compute script

simnh committed:
Merge branch 'features/cli-app' of https://github.com/znes/fuchur into features/cli-app

simnh committed:
Add heat script

gnn committed:
Wrap overly long line

gnn committed:
Drop unsupported Python versions

We're not supporting Python 2.7, as `oemof` decided to not do so and
we're heavily using that.
Python 3.7 is a problem, because `pyproj`, a dependency of
`omoef.tabular`, has to be slightly modified and built from source for
this. Since the goal is to eventually support Python 3.7, I left its
build environments in `tox.ini` but decided to ignore errors on this for
now. This means that builds our build is passing for all supported
versions, but we can still continually monitor the build behaviour in
order to see when it eventually works.
PyPy is Python 2.7 so that gets dropped anyway. Whether PyPy3 works or
not, has to be checked via a continuous integration service, as I don't
have that interpreter at hand.

gnn committed:
Add `oemof.tabular` as a dependency

I noticed this was missing while checking whether the documentation gets
built. Since we are depending on a particular branch of `oemof.tabular`,
the version had to be pinned and the install command had to be changed
in order for dependency links to be followed.

gnn committed:
Update `Development Status`

We are definitely no longer in the planning phase.

gnn committed:
Add the API RST file for `fuchur.scripts`

gnn committed:
Fix `cli` test

gnn committed:
Remove unsupported Python versions from CI

It's not enough to remove them from `tox.ini` if the configuration for
the continuous integration (CI) services still makes them run the
builds.
Duh... stupid me.

gnn committed:
Remove unsupported implementations

Pypy doesn't support `pandas`.

gnn committed:
Limit line length to 79 characters

At least in places where `setup.cfg` allows us to specify that.

gnn committed:
Remove Python 2.7 from `setup.cfg`

Just in case we want to regenerate the whole shebang via `bootstrap.py`.

gnn committed:
Allow Python 3.7 to fail

Hope I specified that correctly enough for Travis to pick it up.

gnn committed:
Change some import sorting options

gnn committed:
Make the style checker happy

This entailed running `black` on all offending files, wrapping some
remaining, overly long lines manually and reordering the imports.
But now it's done and should work again. :)

gnn committed:
Fix `.travis.yml`

Travis rejected it for not being able to parse it.

gnn committed:
Allow Python 3.7 windows CI jobs to fail

gnn committed:
Don't send emails on CI failures

gnn committed:
Make travis be more lenient with timeouts

It's actually only needed for Python 3.4 because that's the only job
that times out, but I didn't find a way to specify it only for that
environment. Shouldn't be a problem for the others though.
See

  https://docs.travis-ci.com/user/common-build-problems/#build-times-out-because-no-output-was-received

for documentation on how this works.

gnn committed:
Fix `appveyor.yml`

gnn committed:
Try fixing `appveyor.yml` differently

I saw the `allow_failures` key was on the same level as `matrix`,
instead of being subordinate to it.
The way it is now should replicate the example. Hope it works.

gnn committed:
Test Travis' patience even more

gnn committed:
Still trying to fix `appveyor.yml`

simnh committed:
Add scenarios

simnh committed:
Update scenarios and move biomass to other script

gnn committed:
Make the linter happy

Two is the number of empty lines you should put after your imports and
the number of empty lines you should put after your imports is two.
Otherwise you will never defeat the Killer Rabbit of Caerbannog. ;)

gnn committed:
Blacken `fuchur/__init__.py`

gnn committed:
Add a dictionary containing built in scenarios

simnh committed:
Move load profile to own function

simnh committed:
Add heat to test.toml

gnn committed:
Allow specifying built in scenarios by name

Now you can specify built in scenarios by their `name` when running
`fuchur construct`.

gnn committed:
Implement hierarchical scenarios

This is done by implementing the `Scenario` class, which has the
following features:

  * It encapsulates loading scenario from a path via the `from_path`
    class method and
  * it looks for a `parents` key in the loaded scenario, which enables
    copying other scenarios into this one.

The `from_path` method does everything that is necessary to load a
scenario from a file. This currently entails:

  * using a special function from `oemof.tabular` instead of just using
    `toml.load` and
  * registering the loaded scenario as known under `fuchur.scenarios`
    under its name and its path.

A scenario's `parents` can be specified as a list of strings, which will
be looked up in `fuchur`'s known scenarios. If they string is not found,
it is tried to be loaded via `Scenario.from_path`.

gnn committed:
Merge improved config handling into features/cli-app

You can now specify builtin scenarios simply by their `name` when
calling `fuchur construct` and you can copy other scenarios into a
scenario by referring to the former via a `parents` entry in the latter.

simnh committed:
Resolve merge conflict

simnh committed:
Add tyndp script

simnh committed:
Add extra script for german dispatchables

Commit:
Fix dispatchable_de

simnh committed:
Exclude DE dispatchable from tyndp

simnh committed:
Remove dispachable de

simnh committed:
Restructure functions, add docstrings, and arguments for generic usage

simnh committed:
Move tyndp to electricity.py, adapt cli and add scenario file

simnh committed:
Remove scenarios file

simnh committed:
Make scenarios work in the way with parents being used as template and update by children

simnh committed:
Add raw data path argument to hydro generation

simnh committed:
Remove obsolete print statement

simnh committed:
Remove moved code from cli for tyndp / nep dispatch model

simnh committed:
Adapt scripts for carrier tech consistency

simnh committed:
Restructure scripts for easier maintainability

simnh committed:
Move nep2019 to electrictiy script

simnh committed:
Update fuchur
